platforms/core-execution/execution/README.md

```diff
@@ -0,0 +1,155 @@
+
+## Optimizations
+
+### Incremental Build (Up-to-Date Checks)
+
+_TBD_
+
+### The Build Cache
+
+_TBD_
+
+### Execution Modes
+
+Note that this differs from _incremental build,_ which involves skipping the entire execution of an up-to-date unit of work.
+
+Fundamentally, the execution engine supports two kinds of work: _incremental_ and _non-incremental_.
+
+- **Non-Incremental Work**: Identified by its full set of inputs.
+  Changing any input effectively creates a new identity.
+- **Incremental Work**: Has some **incremental inputs** that are not part of the work's identity.
+  Changing these incremental inputs does not change the work's identity.
+  However, changing a non-incremental input _does_ change the identity.[^task-identity]
+
+[^task-identity]: This is currently not true for tasks, whose identity is the task's full path.
+The plan is to change this and have tasks execute in workspaces similar to artifact transforms.
+
+Incremental work is executed within the same workspace as long as only incremental inputs are changing.
+This allows the work to reuse outputs from previous executions and update them instead of regenerating everything from scratch.
+To facilitate this, the execution engine provides the action with a list of changes to any incremental input since the last execution in the same workspace.
+
+#### Workspace Allocation
+
+Incremental work is assigned a **mutable workspace**, a directory on disk reused in subsequent builds.
+Execution within these workspaces happens under a lock to prevent multiple units of work from colliding.[^task-locking]
+
+[^task-locking]: For historical reasons, Gradle tasks lack such mutual exclusion, and two separate Gradle daemons running the same task can collide.
+
+In contrast, non-incremental work is assigned a temporary workspace directory upon execution.
+Once execution finishes (or results are loaded from the cache), the workspace directory is atomically moved to an **immutable workspace** that is not modified further.
+
+## Legacy: The Special Case of Gradle Tasks
+
+Some invariants that apply to other units of work do not yet apply to tasks in Gradle.
+This is mostly historical and should be addressed.
+
+### Identity
+
+A task's identity is its full path (e.g., `:project-name:taskName`).
+Crucially, this identity is independent of non-incremental inputs.
+For non-incremental tasks, when a non-incremental input changes, previous outputs are still presented to the action, making it the action's responsibility to clean them up.
+For incremental tasks, the execution engine automatically cleans up outputs when non-incremental inputs change.
+The plan is to extend this behavior to non-incremental tasks as well (or execute them via the same immutable workspaces used for non-incremental transforms).
+However, some tasks do not declare themselves as incremental yet exploit the ability to update their outputs.
```

Comment:
❌ Please clarify in the corresponding section how the engine understands if the unit of work (including tasks) is incremental or not